### PR TITLE
Convert some LoggerMessage.Define to LoggerMessageAttribute

### PR DESCRIPTION
**Pull request by pranavkm**
* [x] DataProtection
* [x] CORS
* [x] Authentication.Core
* [x] HttpOverrides

The updates were performed [using a codefix](https://github.com/pranavkm/LoggerConvert) so the conversions are unlikely to have manual copy-pasta error. Sampling a few to verify the code fix did the right job should suffice.

### CODE REVIEW

File: src/DataProtection/DataProtection/src/Internal/DataProtectionHostedService.cs [hunk 1]
```diff
@@ -33,19 +33,28 @@ public Task StartAsync(CancellationToken token)
...
             }
             catch (Exception ex)
             {
-                // This should be non-fatal, so swallow, log, and allow server startup to continue.
-                // The KeyRingProvider may be able to try again on the first request.
-                _logger.KeyRingFailedToLoadOnStartup(ex);
+                Log.KeyRingFailedToLoadOnStartup(
+                                // This should be non-fatal, so swallow, log, and allow server startup to continue.
+                                // The KeyRingProvider may be able to try again on the first request.
+                                _logger, ex);
```
Comment (pranavkm) [inline src/DataProtection/DataProtection/src/Internal/DataProtectionHostedService.cs:43]:
```suggestion
                // This should be non-fatal, so swallow, log, and allow server startup to continue.
                // The KeyRingProvider may be able to try again on the first request.
                Log.KeyRingFailedToLoadOnStartup(_logger, ex);
```

File: src/DataProtection/DataProtection/src/AuthenticatedEncryption/ManagedAuthenticatedEncryptorFactory.cs [hunk 2]
```diff
@@ -127,5 +127,14 @@ private class AlgorithmActivatorCore<T> : IActivator<T> where T : new()
                 public Func<T> Creator { get; } = Activator.CreateInstance<T>;
             }
         }
+
+        private partial class Log
+        {
+            [LoggerMessage(10, LogLevel.Debug, "Using managed keyed hash algorithm '{FullName}'.", EventName = "UsingManagedKeyedHashAlgorithm")]
+            public static partial void UsingManagedKeyedHashAlgorithm(ILogger logger, string fullName);
+
+            [LoggerMessage(11, LogLevel.Debug, "Using managed symmetric algorithm '{FullName}'.", EventName = "UsingManagedSymmetricAlgorithm")]
```
Comment (Tratcher) [inline src/DataProtection/DataProtection/src/AuthenticatedEncryption/ManagedAuthenticatedEncryptorFactory.cs:136]: This is a good example where everything is using a shared logger and it's hard to ensure the ids are unique. What was the conclusion there? Moving the IDs back to a shared class would be helpful.

Comment (pranavkm) [reply src/DataProtection/DataProtection/src/AuthenticatedEncryption/ManagedAuthenticatedEncryptorFactory.cs:136]: > and it's hard to ensure the ids are unique

The source generator ensures are ids are unique for a given log scope so we think that's good enough. 

In particular, the only time moving the shared one to a scoped one is problematic is when a previously used id is no longer used and consequently can become accidentally re-used in the future. CORS had a couple of these: https://github.com/dotnet/aspnetcore/pull/34699/files#diff-a39be06054f8e5592a940c4591a961448a0f9f0c6c36cd564ba8c0239414f99dR289-R290

Comment (Tratcher) [reply src/DataProtection/DataProtection/src/AuthenticatedEncryption/ManagedAuthenticatedEncryptorFactory.cs:136]: How can it tell these are in the same log scope as adjacent classes? It only references an ILogger here.

File: src/Middleware/HttpOverrides/src/CertificateForwardingFeature.cs [hunk 3]
```diff
@@ -47,5 +47,11 @@ public CertificateForwardingFeature(ILogger logger, StringValues header, Certifi
 
         public Task<X509Certificate2?> GetClientCertificateAsync(CancellationToken cancellationToken)
             => Task.FromResult(ClientCertificate);
+
+        private static partial class Log
+        {
+            [LoggerMessage(0, LogLevel.Warning, "Could not read certificate from header.", EventName = "NoCertificate")]
+            public static partial void NoCertificate(ILogger logger, Exception exception);
```
Comment (Tratcher) [inline src/Middleware/HttpOverrides/src/CertificateForwardingFeature.cs:54]: This is a shared logger, move this back to a shared location.

File: src/Middleware/StaticFiles/src/StaticFileContext.cs [hunk 4]
```diff
@@ -420,5 +417,38 @@ private enum RequestType : byte
             IsGet = 0b_010,
             IsRange = 0b_100,
         }
+
+        private partial class Log
```
Comment (Tratcher) [inline src/Middleware/StaticFiles/src/StaticFileContext.cs:421]: These share a logger with the middleware and should live in a shared location.

File: src/DataProtection/DataProtection/src/AuthenticatedEncryption/ManagedAuthenticatedEncryptorFactory.cs (hunk 2 above)
Comment (Tratcher) [reply src/DataProtection/DataProtection/src/AuthenticatedEncryption/ManagedAuthenticatedEncryptorFactory.cs:136]: I was wrong about this instance being shared, but I've found a few more shared that I'll flag.

File: src/Security/Authentication/Core/src/RemoteAuthenticationHandler.cs [hunk 5]
```diff
@@ -106,7 +106,7 @@ public virtual async Task<bool> HandleRequestAsync()
 
             if (exception != null)
             {
-                Logger.RemoteAuthenticationError(exception.Message);
+                Log.RemoteAuthenticationError(Logger, exception.Message);
```
Comment (Tratcher) [inline src/Security/Authentication/Core/src/RemoteAuthenticationHandler.cs:109]: These are also using a shared logger with the base class.

**Comment by JamesNK** [comment]
I'm really not a fan of splitting up previously shared logger like this. Having these log messages split across types, each with different IDs (e.g. why is this ID 3 and the next ID 5?), doesn't seem to add any value.

I think for shared logging types like this it is simpler to just adopt the source generator but keep all the log messages on one type.

File: src/DataProtection/DataProtection/src/AuthenticatedEncryption/ManagedAuthenticatedEncryptorFactory.cs (hunk 2 above)
Comment (JamesNK) [reply src/DataProtection/DataProtection/src/AuthenticatedEncryption/ManagedAuthenticatedEncryptorFactory.cs:136]: Can we just leave all the logs that were in one file together? It just seems simpler and less confusing than to force a new pattern afterwards.

**Comment by pranavkm** [comment]
> What number do I use if I add a new message?

Outside of the ones where the logger is shared across multiple types (for which we'll commonize again), picking any id should be fine, no? 

If we trust that the tool correctly does the right job copying the log statements around, wouldn't it be preferable to have a consistent pattern across the repo?

**Comment by pranavkm** [comment]
How about this? I'll change this PR to only update the LoggerMessage.Define -> LoggerMessageAttribute. I can do separate follow ups to switch it to use a nested Log class. That would be less traumatic since it would largely be moving methods around and would not require any scrutiny.

**Comment by Tratcher** [comment]
Fair, most of the concern is with the second part of that and your real goal is the first part.

**Pull request closed without merging**